onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,695 @@
+            }
+
+            auto axisIdx = OperatorHelper::HandleNegativeAxis(static_cast<int32_t>(axis), rank);
+
+            // if (context->IsInputValid(1))
+            // {
+            //     // If dft_length is specified, then we should honor the shape.
+            //     // Set the output dimension to match the dft_length on the axis.
+            //     // If onesided this will be adjusted later on.
+            //     // However, this is not supported as of now, so do nothing...
+            // }
+
+            // In general the output shape will match the input shape exactly
+            // So initialize the output shape with the input shape
+            std::vector<uint32_t> inputDims(rank);
+            ORT_THROW_IF_FAILED(context->GetInputTensorShape(0, rank, inputDims.data()));
+            auto outputDims = inputDims;
+            outputDims.back() = 2;
+
+            // When DFT is onesided, the output shape is half the size of the input shape
+            // along the specified axis.
+            if (isOnesided) {
+                auto axisDimension = outputDims.at(axisIdx);
+                // We need to update the output shape dimension along the specified axis,
+                // but sometimes the dimension will be a free dimension or be otherwise unset.
+                // Only perform inference when a input dimension value exists.
+                auto originalSignalSize = axisDimension;
+                auto halfSignalSize = (originalSignalSize >> 1) + 1;
+                outputDims.at(axisIdx) = halfSignalSize;
+            }
+
+            ORT_THROW_IF_FAILED(context->SetOutputTensorShape(0, rank, outputDims.data()));
+        }
+        catch (...)
+        {
+            return E_FAIL;
+        }
+
+        return S_OK;
+    }
+};
+
+class GpuDFTOperatorFactory : public WRL::Base<IMLOperatorKernelFactory>
+{
+public:
+    STDMETHOD(CreateKernel)(
+        IMLOperatorKernelCreationContext* context,
+        IMLOperatorKernel** kernel)
+    {
+        try
+        {
+            auto dftOperator = wil::MakeOrThrow<GpuDFTOperator>(context);
+            dftOperator.CopyTo(kernel);
+            return S_OK;
+        }
+        catch (...)
+        {
+            return E_FAIL;
+        }
+    }
+
+    static void RegisterDFTKernel(IMLOperatorRegistry* registry)
+    {
+        MLOperatorKernelDescription kernelDescription = {};
+        kernelDescription.domain = "";
+        kernelDescription.name = "DFT";
+        kernelDescription.minimumOperatorSetVersion = 17;
+        kernelDescription.executionType = MLOperatorExecutionType::D3D12;
+
+        // T1: tensor(float16), tensor(float), tensor(double), tensor(bfloat16)
+        MLOperatorEdgeTypeConstrant t1Constraint;
+        t1Constraint.typeLabel = "T1";
+        std::vector<MLOperatorEdgeDescription> t1AllowedEdges
+        {
+            //MLOperatorEdgeDescription { MLOperatorEdgeType::Tensor, (uint64_t)MLOperatorTensorDataType::Float16 },
```

Comment:
>MLOperatorTensorDataType::Float16 [](http://example.com/codeflow?start=80&length=33)

Is float16 in mind for the future? AFAIK all our DML EP kernels that support float32 also support float16.

Reply:
I think so, but im not sure how to enable fp16 here. Should have a chat.

Reply:
When promoted to DML we should make the associated changes.

Reply:
Okie dokie.